### ISSUE TITLE
Proper detection of offline server

### ISSUE THREAD

**Issue by GreenArrow18** [colyseus/colyseus-defold#37]
Hi @endel,

I am trying to reliably capture an offline colyseus server, so a proper message can be displayed on the game client. The client doesn't always respond with an "offline" message though, and will sometimes throw a json error.

I just tried the example in the repository. The default properly gives an offline message (no colyseus server running):
`client = Colyseus.new("ws://localhost:2567")`

However, once you change it to something like:
`client = Colyseus.new("wss://echo.websocket.org")`

It throws a script error instead:
```
ERROR:SCRIPT: /colyseus/client.lua:132: Invalid JSON primitive: <html>
stack traceback:
	[C]: in function 'decode'
	/colyseus/client.lua:132: in function </colyseus/client.lua:131>
```

The example above is responding with a 404 error for instance, but the json.decode call throws a runtime error before the response status can even be checked. I was looking at the codes, but not so sure how to properly fix it...

**Comment by endel** [comment]
Hi @GreenArrow18, 

Whenever you call any of the matchmaking functions (`.join_or_create()`, etc.), the error callback should be triggered immediately if the server is not accessible. Have you tried handling the error like in the example below?

```lua
colyseus = Colyseus.new(endpoint, false)

colyseus:join_or_create("raft", {
	debug = global.isDebug,
	tutorial = true
}, function(err, room)
	if err then
		-- error: server offline?
		pprint(err)
		return
	end

	-- handle success
end)
```

**A quick note:** the `echo.websocket.org ` is a plain WebSocket server, and won't work properly with the Colyseus client because of the way Colyseus expects messages to come using a particular format.

**Comment by GreenArrow18** [comment]
Hi @endel,

Yes, that's how things are set-up on my end.

I just made the websocket.org example because the runtime error is easily reproducible using that one in the repository example. However, it also happens in my own server. 

On my first connection attempt, I would get an offline message properly. The response table is this:
```
{ --[[00000239BFD84B60]]
  status = 0,
  response = "",
  headers = { } --[[00000239BFD85F50]]
}
```

However, when I try to connect again, I would now get a 502 error from nginx, and the colyseus client is unable to parse the message properly. Similar to connecting to websocket.org server, except that returns a 404 error.

Here's the actual response table I get:
```
{ --[[000001FBC02E61A0]]
  status = 502,
  response = "<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.19.0</center>
</body>
</html>
",
  headers = { --[[000001FBC02E62F0]]
    server = "nginx/1.19.0",
    content-type = "text/html",
    connection = "keep-alive",
    content-length = "157",
    strict-transport-security = "max-age=15768000; includeSubDomains; preload",
    date = "Thu, 28 Jan 2021 17:32:45 GMT"
  }
}
```

NOTES: I am using nginx. Colyseus server was purposely offline.

I understand the websocket.org server is not compatible with colyseus, but my thought is the client should reliably parse the initial connect response codes, and not throw a runtime error?

The runtime error on client.lua happens on the call to json.decode:
```
local data = response.response ~= '' and json.decode(response.response)
local has_error = (response.status >= 400)
```

Perhaps check for has_error, before decoding the data? But you have other checks below that code, so I didn't want to mess things up.

Here's what confuses me more. Note that I have a connect time out of 3 seconds only.
1. First connection attempt: I will get a response code of 0, so offline message is properly detected.
2. Succeeding connection attempts: I will get a response code of 502 from nginx, which the colyseus client can not parse properly. However this 502 response code is only returned if the 2nd connection attempt was made approx **within 10 seconds**. If I wait for at least 10 seconds before attempting another connect, then I get the usual response code of 0 so the colyseus client will once again return an offline message properly.

Perhaps you can reproduce on your end? Try a second connection attempt to an offline colyseus server within a span of less than 10 sec?

I am using Windows 10, and I can reproduce this in the editor build all the time. I am not sure if it's my nginx setup only, or somehow a Defold socket issue, but if the colyseus client can parse the response codes properly I think the issue will be resolved.

NOTE: I am still using the old Defold websocket. I am tracking another potential issue in the new websocket, which may or may not be related with the issue discussed here.

**Comment by britzl** [comment]
> I am tracking another potential issue in the new websocket, which may or may not be related with the issue discussed here.

Can you share some additional information regarding this? Anything we can help you with?

**Comment by GreenArrow18** [comment]
Hi @endel,

To fix this, I changed the following:
```
local data = response.response ~= '' and json.decode(response.response)
local has_error = (response.status >= 400)
...
```

To this:
```
local has_error = (response.status >= 400)
if (has_error) then
  return callback("offline")
end
local data = response.response ~= '' and json.decode(response.response)
...
```

Anything with a response code of >=400 most likely didn't reach the colyseus server anyway, so I just avoid parsing the non-valid json data from nginx (it returns an html type in my setup). I am not making a PR because this doesn't try to decode the actual error at all. It's just a quick fix that works well for my case, so I am closing the issue now.

**Comment by GreenArrow18** [comment]
> Can you share some additional information regarding this? Anything we can help you with?

Hi @britzl,

I was testing this same issue back in December using the new websocket extension, and then I was seeing additional websocket messages in the console, like the websocket is already closed or something (on succeeding connection attempts). I had to revert back to the old websocket because of that.

I just tried with the latest websocket extension version though, along with the latest Defold, and I can't seem to reproduce the error anymore, so all good now!

I'll let you know if I encounter some issues. Thanks!
